1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,99 @@
+---
+RFC: RFCnnnn
+Author: Paul Higinbotham
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine
+Comments Due: July 18, 2019
+Plan to implement: Yes
+---
+
+# PowerShell ForEach-Object -Parallel Cmdlet
+
+This RFC proposes a new parameter set for the existing ForEach-Object cmdlet to parallelize script block executions, instead of running them sequentially as it does now.  
+
+## Motivation
+
+    As a PowerShell User,
+    I can do simple fan-out concurrency with the PowerShell ForEach-Object cmdlet, without having to obtain and load a separate module, or deal with PowerShell jobs unless I want to.
+
+## Specification
+
+There will be two new parameter sets added to the existing ForeEach-Object cmdlet to support both synchronous and asynchronous operations for parallel script block execution.
+For the synchronous case, the `ForEach-Object` cmdlet will not return until all parallel executions complete.
+For the asynchronous case, the `ForEach-Object` cmdlet will immediately return a PowerShell job object that contains child jobs of each parallel execution.
+
+### Implementation details

```

Comment:
Will ForEach-Object implement pipeline backpressure?  If not, how will memory usage for pipelines like `Get-Content giantfile.txt | ForEach-Object -Parallel {'CPU-bound operation'}` be limited?  Is so, how will backpressure work?

Reply:
I think the default `ThrottleLimit` would be the way to limit how many parallel jobs will run, no?

Reply:
> think the default ThrottleLimit would be the way to limit how many parallel jobs will run, no?

@daxian-dbw That's what I understand.  The concern I'm raising here isn't about the number of parallel jobs.  Rather, it is about how much of the output from `Get-Content giantfile.txt` is read into the input buffer of `ForEach-Object -Parallel` at any one time.  The way I read the RFC currently, the behavior that respect is not specified.  If no backpressure is implemented, then the amount of memory that could be consumed by `ForEach-Object -Parallel` is unbounded.  If backpressure is implemented, then the particularities of that behavior will affect how performant `ForEach-Object -Parallel` is under different conditions.